File: advent-of-code/src/main/java/com/willmolloy/adventofcode/common/grid/CharGrid.java
package com.willmolloy.adventofcode.common.grid;

import static com.google.common.base.Preconditions.checkNotNull;

import java.util.Arrays;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

/**
 * 2D char grid.
 *
 * @author <a href=https://willmolloy.com>Will Molloy</a>
 */
public class CharGrid implements BaseGrid<Character, CharGrid, MutableCharGrid> {
  private final char[][] array;

  public CharGrid(char[][] array) {
    this.array = checkNotNull(array);
  }

  @Override
  public int height() {
    return array.length;
  }

  @Override
  public int width() {
    return array[0].length;
  }

  @Override
  public int size() {
    return height() * width();
  }

  @Override
  public Stream<Point> points() {
    return IntStream.range(0, height())
        .mapToObj(y -> IntStream.range(0, width()).mapToObj(x -> new Point(x, y)))
        .flatMap(Function.identity());
  }

  @Override
  public boolean inbounds(Point point) {
    // check height before width, potential IOOB
    return point.y() >= 0 && point.y() < height() && point.x() >= 0 && point.x() < width();
  }

  @Override
  public MutableCharGrid toMutable() {
    char[][] clone = Arrays.stream(array).map(char[]::clone).toArray(char[][]::new);
    return new MutableCharGrid(clone);
  }

  /** Get value at a given point. */
  public char get(Point point) throws ArrayIndexOutOfBoundsException {
    return array[(int) point.y()][(int) point.x()];
  }

  @Override
  public final String toString() {
    return Arrays.stream(array).map(String::new).collect(Collectors.joining("\n"));
  }
}

File: advent-of-code/src/main/kotlin/com/willmolloy/adventofcode/common/grid/CharGrid.kt
package com.willmolloy.adventofcode.common.grid

/**
 * 2D char grid.
 *
 * @author <a href=https://willmolloy.com>Will Molloy</a>
 */
open class CharGrid(private val array: Array<CharArray>) : Grid<Char, CharGrid, MutableCharGrid> {

  override fun height(): Int = array.size

  override fun width(): Int = array[0].size

  override fun size(): Int = height() * width()

  override fun points(): Sequence<Point> = sequence {
    for (y in 0 until height()) {
      for (x in 0 until width()) {
        yield(Point(x, y))
      }
    }
  }

  // check height before width, potential IOOB
  override fun inbounds(point: Point): Boolean =
    point.y in 0 until height() && point.x in 0 until width()

  override fun toMutable(): MutableCharGrid {
    val clone = array.map { it.clone() }.toTypedArray()
    return MutableCharGrid(clone)
  }

  /** Get value at a given point. */
  @Throws(ArrayIndexOutOfBoundsException::class)
  operator fun get(point: Point): Char = array[point.y.toInt()][point.x.toInt()]

  @Throws(ArrayIndexOutOfBoundsException::class)
  operator fun get(row: Int, col: Int): Char = array[row][col]

  override fun toString(): String = array.joinToString("\n") { String(it) }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as CharGrid

    return array.contentDeepEquals(other.array)
  }

  override fun hashCode(): Int {
    return array.contentDeepHashCode()
  }
}
